<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="prev" title="General Usage" href="usage.html" />

    <!-- Generated with Sphinx 7.0.1 and Furo 2023.05.20 -->
        <title>Internal Functions - chromTools 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=e6660623a769aa55fea372102b9bf3151b292993" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">chromTools 0.1 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">chromTools 0.1 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">General Usage</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Internal Functions</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="internal-functions">
<h1>Internal Functions<a class="headerlink" href="#internal-functions" title="Permalink to this heading">#</a></h1>
<section id="chromtools-complete">
<h2>chromTools complete<a class="headerlink" href="#chromtools-complete" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="chromTools.chmm_cmd.make_binary_data_from_bed">
<span class="sig-prename descclassname"><span class="pre">chromTools.chmm_cmd.</span></span><span class="sig-name descname"><span class="pre">make_binary_data_from_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gridcontrol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sumgridcontrol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpresentcontrol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/chromTools/chmm_cmd.html#make_binary_data_from_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#chromTools.chmm_cmd.make_binary_data_from_bed" title="Permalink to this definition">#</a></dt>
<dd><p>Binarize BED data, both directly and with control.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em>) – File number.</p></li>
<li><p><strong>options</strong> (<em>object</em>) – Namespace object.</p></li>
<li><p><strong>control</strong> (<em>bool</em><em> or </em><em>list</em>) – If not False (default), control contains a list of control files.</p></li>
<li><p><strong>szchromfile</strong> (<em>str</em>) – The name of the file containing chromosome information. It is a two-column file with the first column representing the chromosome and the second column representing the chromosome length.</p></li>
<li><p><strong>szcontroldir</strong> (<em>str</em>) – The directory containing the bed files with the control data. If set to None, no control data is used.</p></li>
<li><p><strong>szchromlengthfile</strong> (<em>str</em>) – The name of the two-column file containing chromosome and strand information.</p></li>
<li><p><strong>szmarkdir</strong> (<em>str</em>) – The directory containing the bed files with the regular mark data.</p></li>
<li><p><strong>nflankwidthcontrol</strong> (<em>int</em>) – Specifies the number of bins used in both directions to estimate the background. This attribute is only relevant if control data is being used.</p></li>
<li><p><strong>nshift</strong> (<em>int</em>) – The number of bases a read should be shifted in the 5’ to 3’ direction of a read.</p></li>
<li><p><strong>bcenterinterval</strong> (<em>bool</em>) – If True, the center of the read is used instead of shifting if the read has already been extended.</p></li>
<li><p><strong>noffsetleft</strong> (<em>int</em>) – The amount that should be subtracted from the left coordinate to make it 0-based inclusive.</p></li>
<li><p><strong>noffsetright</strong> (<em>int</em>) – The amount that should be subtracted from the right coordinate to make it 0-based inclusive.</p></li>
<li><p><strong>szoutputsignaldir</strong> (<em>str</em>) – If not None, the intermediate signal data will be printed to this directory.</p></li>
<li><p><strong>szoutputbinarydir</strong> (<em>str</em>) – The directory where the binarized data will be printed.</p></li>
<li><p><strong>szoutputcontroldir</strong> (<em>str</em>) – If not None, the intermediate control signal data will be printed to this directory.</p></li>
<li><p><strong>dpoissonthresh</strong> (<em>float</em>) – The tail probability threshold on the Poisson distribution.</p></li>
<li><p><strong>dfoldthresh</strong> (<em>float</em>) – The fold threshold required for a present call.</p></li>
<li><p><strong>bcontainsthresh</strong> (<em>bool</em>) – If True, the Poisson cutoff should be the highest value that still contains the dpoissonthresh probability. If False, it requires strictly greater.</p></li>
<li><p><strong>npseudocountcontrol</strong> (<em>int</em>) – An integer pseudocount that is uniformly added to every interval to smooth the control data.</p></li>
<li><p><strong>nbinsize</strong> (<em>int</em>) – The number of base pairs in a bin</p></li>
<li><p><strong>szcolfields</strong> (<em>str</em>) – A comma-delimited string indicating the 0-based columns of the chromosome, start, end,
and optionally strand position. If set to None, the default values are used (0, 1, 2) for
chromosome, start, and end, with the strand as the sixth column or the last column if fewer
columns are present.</p></li>
<li><p><strong>dcountthresh</strong> (<em>float</em>) – The absolute signal threshold for a present call.</p></li>
<li><p><strong>bbinarizebam</strong> (<em>bool</em>) – If True, reads files as BAM files; otherwise, reads files as BED files.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – Invalid line found in the chromosome length file if fewer than two cols found</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A tuple containing the count of present marks across included chromosomes and the total number of bins
across included chromosomes.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple[int, int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="chromTools.chmm_cmd.load_grid">
<span class="sig-prename descclassname"><span class="pre">chromTools.chmm_cmd.</span></span><span class="sig-name descname"><span class="pre">load_grid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpresent</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpresentmarks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nshift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nbinsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noffsetleft</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noffsetright</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hmfiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">szcell</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">szmarkdir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hmchrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ninitval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bcontrol</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/chromTools/chmm_cmd.html#load_grid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#chromTools.chmm_cmd.load_grid" title="Permalink to this definition">#</a></dt>
<dd><p>Converts read level data for a cell into integer count information.</p>
<p>This function populates the provided grid with data obtained from mark files. It iterates through mark files for each
cell type and mark, calculates the bin index for each data point, and updates the corresponding grid cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grid</strong> (<em>numpy.ndarray</em>) – The array to load the read counts into.</p></li>
<li><p><strong>bpresent</strong> (<em>list</em>) – Indicates if there is a read for a chromosome with an index in hmchrom.</p></li>
<li><p><strong>bpresentmark</strong> (<em>list</em>) – Indicates if the mark is present in the cell type.</p></li>
<li><p><strong>marks</strong> (<em>list</em>) – Contains the names of the header marks.</p></li>
<li><p><strong>nshift</strong> (<em>int</em>) – The number of bases a read should be shifted in the 5’ to 3’ direction of a read.</p></li>
<li><p><strong>nbinsize</strong> (<em>int</em>) – The number of base pairs in a bin.</p></li>
<li><p><strong>noffsetleft</strong> (<em>int</em>) – The amount that should be subtracted from the left coordinate so it is 0-based inclusive.</p></li>
<li><p><strong>noffsetright</strong> (<em>int</em>) – The amount that should be subtracted from the right coordinate so it is 0-based inclusive.</p></li>
<li><p><strong>hmfiles</strong> (<em>dict</em>) – Maps cell and mark to an actual read file.</p></li>
<li><p><strong>szcell</strong> (<em>str</em>) – The cell we are interested in.</p></li>
<li><p><strong>szmarkdir</strong> (<em>str</em>) – The directory with bedfiles to read.</p></li>
<li><p><strong>hmchrom</strong> (<em>dict</em>) – Maps chromosome names to an index.</p></li>
<li><p><strong>ninitval</strong> (<em>int</em>) – The value that the data should be initialized to.</p></li>
<li><p><strong>bcontrol</strong> (<em>bool</em>) – if True, data is control dataz</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>grid (numpy.ndarray): The updated grid with the read counts.</p>
</dd>
</dl>
<dl class="simple">
<dt>Calculation Details:</dt><dd><ul class="simple">
<li><p>Initializes necessary variables and data structures.</p></li>
<li><p>Iterates over each mark file for each cell type.</p></li>
<li><p>Checks for the presence of control data or missing data and handles accordingly.</p></li>
<li><p>Retrieves mark file data, extracts relevant information, and calculates the bin index.</p></li>
<li><p>Updates the corresponding grid cell with the calculated index and mark information.</p></li>
<li><p>Sets the presence flags for chromosomes and marks based on the loaded data.</p></li>
</ul>
</dd>
<dt>Raises:</dt><dd><p>ValueError: If the column index exceeds the maximum index for the mark file data.
ValueError: If an invalid strand is encountered in the mark file data.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="chromTools.chmm_cmd.determine_mark_thresholds_from_binned_data_array">
<span class="sig-prename descclassname"><span class="pre">chromTools.chmm_cmd.</span></span><span class="sig-name descname"><span class="pre">determine_mark_thresholds_from_binned_data_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpresent</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpoissonthresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dfoldthresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bcontainsthresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dcountthresh</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/chromTools/chmm_cmd.html#determine_mark_thresholds_from_binned_data_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#chromTools.chmm_cmd.determine_mark_thresholds_from_binned_data_array" title="Permalink to this definition">#</a></dt>
<dd><p>Determines the Poisson cutoffs based on the provided data.</p>
<p>This function calculates the thresholds for each mark based on the provided binned data array. It uses the Poisson
distribution and other parameters to determine the cutoffs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grid</strong> (<em>numpy.ndarray</em>) – The integer data values from which to determine the Poisson cutoffs.</p></li>
<li><p><strong>bpresent</strong> (<em>numpy.ndarray</em>) – A vector indicating which indices of ‘grid’ to include in the analysis.</p></li>
<li><p><strong>dpoissonthresh</strong> (<em>float</em>) – The tail probability threshold on the Poisson distribution.</p></li>
<li><p><strong>dfoldthresh</strong> (<em>float</em>) – The fold threshold required for a present call.</p></li>
<li><p><strong>bcontainsthresh</strong> (<em>boolean</em>) – If True, the Poisson cutoff should be the highest value that still contains the ‘dpoissonthresh’ probability.
If False, it requires strictly greater.</p></li>
<li><p><strong>dcountthresh</strong> (<em>float</em>) – The absolute signal threshold for a present call.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>thresholds (list): Each element in this list represents the Poisson cutoff for a specific mark.</p>
</dd>
</dl>
<dl class="simple">
<dt>Calculation Details:</dt><dd><ul class="simple">
<li><p>Initializes variables and data structures.</p></li>
<li><p>Computes the sum of data values for each mark across all relevant bins and chromosomes.</p></li>
<li><p>Calculates the total number of bins considered for threshold determination.</p></li>
<li><p>Computes the threshold for each mark based on the Poisson distribution.</p></li>
<li><p>Adjusts the threshold based on the bcontainsthresh parameter.</p></li>
<li><p>Sets the final threshold for each mark as the maximum value among various calculations.</p></li>
</ul>
</dd>
<dt>Note:</dt><dd><ul class="simple">
<li><p>The thresholds are computed based on the Poisson distribution and various parameters provided.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="chromTools.chmm_cmd.determine_mark_thresholds_from_binned_data_array_against_control">
<span class="sig-prename descclassname"><span class="pre">chromTools.chmm_cmd.</span></span><span class="sig-name descname"><span class="pre">determine_mark_thresholds_from_binned_data_array_against_control</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gridcontrol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpresent</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bpresentcontrol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpoissonthresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dfoldthresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bcontainsthresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dcountthresh</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/chromTools/chmm_cmd.html#determine_mark_thresholds_from_binned_data_array_against_control"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#chromTools.chmm_cmd.determine_mark_thresholds_from_binned_data_array_against_control" title="Permalink to this definition">#</a></dt>
<dd><p>Determines the Poisson cutoffs based on the provided data.</p>
<p>This function calculates the thresholds for each mark by comparing the binned data array against the control data.
It uses the Poisson distribution to determine the thresholds based on various parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grid</strong> (<em>numpy.ndarray</em>) – The integer data values from which to determine the Poisson cutoffs.</p></li>
<li><p><strong>gridcontrol</strong> (<em>numpy.ndarray</em>) – The control data to which the thresholds will be relative.</p></li>
<li><p><strong>bpresent</strong> (<em>numpy.ndarray</em>) – A vector indicating which indices of ‘grid’ to include in the analysis.</p></li>
<li><p><strong>bpresentcontrol</strong> (<em>numpy.ndarray</em>) – A vector indicating which indices of ‘gridcontrol’ to include in the analysis.</p></li>
<li><p><strong>dpoissonthresh</strong> (<em>float</em>) – The tail probability threshold on the Poisson distribution.</p></li>
<li><p><strong>dfoldthresh</strong> (<em>float</em>) – The fold threshold required for a present call.</p></li>
<li><p><strong>bcontainsthresh</strong> (<em>boolean</em>) – If True, the Poisson cutoff should be the highest value that still contains the ‘dpoissonthresh’ probability.
If False, it requires strictly greater.</p></li>
<li><p><strong>dcountthresh</strong> (<em>float</em>) – The absolute signal threshold for a present call.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>thresholds (list): Each element in this list represents the Poisson cutoff for a specific mark.</p>
</dd>
</dl>
<dl class="simple">
<dt>Calculation Details:</dt><dd><ul class="simple">
<li><p>Initializes variables and data structures.</p></li>
<li><p>Computes the total number of reads for each mark and its matched control.</p></li>
<li><p>Determines the maximum control value found for each mark.</p></li>
<li><p>Determines the background control values encountered for each mark.</p></li>
<li><p>Calculates thresholds for each mark based on the observed data and control values.</p></li>
</ul>
</dd>
<dt>Note:</dt><dd><ul class="simple">
<li><p>The thresholds are computed based on the Poisson distribution and various parameters provided.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="usage.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">General Usage</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Jessica Shields
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Internal Functions</a><ul>
<li><a class="reference internal" href="#chromtools-complete">chromTools complete</a><ul>
<li><a class="reference internal" href="#chromTools.chmm_cmd.make_binary_data_from_bed"><code class="docutils literal notranslate"><span class="pre">make_binary_data_from_bed()</span></code></a></li>
<li><a class="reference internal" href="#chromTools.chmm_cmd.load_grid"><code class="docutils literal notranslate"><span class="pre">load_grid()</span></code></a></li>
<li><a class="reference internal" href="#chromTools.chmm_cmd.determine_mark_thresholds_from_binned_data_array"><code class="docutils literal notranslate"><span class="pre">determine_mark_thresholds_from_binned_data_array()</span></code></a></li>
<li><a class="reference internal" href="#chromTools.chmm_cmd.determine_mark_thresholds_from_binned_data_array_against_control"><code class="docutils literal notranslate"><span class="pre">determine_mark_thresholds_from_binned_data_array_against_control()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>